### ISSUE TITLE
Want variable string manipulation

### ISSUE THREAD

**Issue by kballard** [fish-shell/fish-shell#156]
I'd really like to see a way to do basic string manipulation of variables, similar to how bash provides substring offsets, pruning of prefix/suffix patterns, and even regular expressions. Having to spawn a separate program just to dice up one string is rather slow and awkward.

**Comment by coledot** [comment]
+1. This could also be applied as a solution to #828 in that `alias` would no longer need to fork/exec on every single call.

**Comment by brandonweiss** [comment]
:+1:

**Comment by alphapapa** [comment]
By the way, what about regexp matching?  What about a `(match|replace) -r` to use regexps instead of globs?

**Comment by frangio** [comment]
What's the status on this?

**Comment by ridiculousfish** [comment]
I haven't done any work on this yet. kballard's proposal looks to be the most complete by far - what do you think of it?

**Comment by frangio** [comment]
I think overall @kballard's interface is ok and it fits fish's design philosophy quite well. However, it would be nice to get some more functionality out of it.

I'm in favour of a separate `regex` builtin, but maybe we can squeeze some more pattern substitution capabilities into the `string replace` subcommand. What I have in mind is something similar to GNU Make's `patsubst` function [[1]](https://www.gnu.org/software/make/manual/html_node/Text-Functions.html). For example, `string replace "x/*.a" "y/*.b" x/test.a` would result in the string `y/test.b`. This is basically the same as `patsubst` but using the asterisk instead of the percent sign. What do you think of this?

On a different note, it would be nice if the `string` and `regex` subcommands received any number of string arguments, performed the given operation on all of them, and returned the results in an array.

**Comment by lloeki** [comment]
LGTM.

Regarding @frangio's last note, and certainly another issue entirely (although it'd impact this one in some way), it may be the opportunity to evaluate the need for a `map` builtin, generalising the extremely frequent array->array transformations by combining `for` and capture. Sort of example:

```
xa = ( "x/test.a" "x/other.a" )
yb = map i in $xa
  string replace "x/*.a" "y/*.b" $i
end
```

Output of whatever is in the map is captured on each iteration. Can include multiple commands, even non-builtins. Same semantics as `()` capture means no fork needed to capture builtins. Don't want to capture a command's output? `> /dev/null`.

**Comment by kballard** [comment]
I had intended to try and do this work myself, but I never found the time. I'm still far too busy to give this the attention it deserves, but if @ridiculousfish or someone else finds the time that would be excellent.

As for regular expression support, I agree that it would be nice. I'm inclined to say this should be done as a separate `--regex` flag to `string match` and `string replace`. I'm very strongly against trying to come up with our own syntax for this, I'd rather just use a pre-existing syntax (and, presumably, engine). A few options here:

* POSIX.2 "basic" regular expressions. I think this is a terrible idea, but it's the most obvious one.
* POSIX.2 "extended" regular expressions. Not a fan of this either. It's better than basic, but there's plenty of better options.
* ICU regular expressions. I'm not entirely certain what capabilities these have, but it's the regular expression engine that Cocoa uses, and it presumably has full unicode support. I believe it's reasonably full-featured.
* PCRE
* Oniguruma
* [Onigmo](https://github.com/k-takata/Onigmo), which is apparently an enhanced fork of Oniguruma

I'm sure there's other reasonable options too, but that's what I can think of off the top of my head.

**Comment by ridiculousfish** [comment]
We should also figure out how the regexp syntax relates to fish syntax, e.g. how many backslashes it takes to match a literal backslash. Once you start having to escape your escapes, things get confusing fast.

**Comment by frangio** [comment]
I agree about `--regex` being a flag for the `string` subcommands, and about using a pre-existing engine for the regular expression support. I personally prefer PCRE-compatible syntax.

The escaping shouldn't be a problem since fish escape sequences aren't expanded within quotes, right?

As for the proposed `map` builtin, I don't think I've seen elsewhere in fish blocks that return values, and I would prefer to keep those semantics. This does deserve to be discussed in a different issue, but what I think would be nice is to keep most builtins "array enabled", performing an implicit map over its arguments. (To accomplish a similar thing for commands in general, it should be enough to give `set` the ability of appending to an array (i.e. an `-a`/`--append` option), and then combine this with a for loop.)

**Comment by ridiculousfish** [comment]
[This link](http://www.regular-expressions.info/refbasic.html) has a breakdown of different regex syntaxes. You can select them from the dropdown to compare. It looks like one of the significant differences between PCRE and POSIX ERE is that PCRE supports additional escapes, like `\7` to match the bell character. Those could get ugly fast.

An example of the escaping madness: to match a literal backslash, `egrep '\\\\'`. To match a literal dollar sign using double quotes is `egrep "\\\$"`. That's the sort of stuff I can never get right on the first try. I don't know what the solution is, but it feels like there's an opportunity to really solve that problem.

**Comment by alphapapa** [comment]
I'm not sure if this comment is more than noise, but I'm in favor of POSIX EREs, because those seem to be the most familiar in shell scripts.  For example, Bash's `=~` operator uses EREs.

Basically, I think the Bash `[[ $string =~ pattern ]]` syntax is excellent and underappreciated.  I still see so many people writing Bash (not `/bin/sh`) scripts using single-bracket tests, piping through `grep`, etc.

Personally, I'd vote strongly against using an unusual regexp library or syntax.  It's great that Fish can fix some of the shell idiosyncrasies that have existed for a long time, but regexps are like their own problem space, and I think Fish should go with the flow here.  Regexps are enough of a challenge without having to learn a new type.  Perhaps in time, Fish could add its own, improved regexp type in addition to the standard ones.

**Comment by ridiculousfish** [comment]
It looks like bash is getting some mileage out of its `[[` special-syntax here. For example:

    [[ aaab =~ a*b ]] && echo matched

It's not necessary to escape the asterisk because wildcard expansion is not performed within double square brackets.

**Comment by alphapapa** [comment]
Yes, more specifically, from the man page:

>Any part of the pattern may be quoted to force the quoted portion to be matched as a string.  Bracket expressions in regular expressions must be treated  carefully, since normal quoting characters lose their meanings between brackets.  If the pattern is stored in a shell variable, quoting the variable expansion forces the entire pattern to be matched as a string.

It's also not necessary to quote variables within double-brackets (e.g. on the left side of the operator, if there is an operator).  The bottom line is that, whenever writing Bash-specific scripts, one should always use double-brackets instead of single-brackets.

**Comment by lloeki** [comment]
Regarding regex syntax, the [Go regexp page][0] has a rationale on their choice (such as consistency with expectations coming from [other languages][1] and [linear time][2]). This might shed some light into what choice of syntax to make.

[0]: http://golang.org/pkg/regexp/
[1]: http://code.google.com/p/re2/wiki/Syntax
[2]: http://swtch.com/~rsc/regexp/regexp1.html

**Comment by twlz0ne** [comment]
+1

**Comment by crmne** [comment]
+1!

**Comment by aboisvert** [comment]
+1 to adding string manipulation as built-ins.   Love all the participation on this thread.

**Comment by msteed** [comment]
I'd like to take this on. I can't promise to get it all in the first pass but I'll do what I can.

**Comment by msteed** [comment]
https://github.com/msteed/fish-shell/tree/string-builtin

I have a couple commands implemented (`string length` and the non-regex `string match`). I'll keep working on it as time allows. Feedback & contributions are welcome.

**Comment by msteed** [comment]
`string split` and `string join` are done

**Comment by msteed** [comment]
`string sub` is done (but lacks tests).

The hardest implementation task left is the regex functionality. I'd like some direction from @ridiculousfish or another project decision maker on what engine to use and how to go about incorporating it into fish. Since fish is GPL2, do we just take a GPL2-licensed engine and add it to the fish repository?

**Comment by ridiculousfish** [comment]
@msteed it's great you've opted to take this on! You may want to propose the interface (i.e. commands) sooner rather than later, so that if we have feedback you don't end up wasting work or get too attached to an implementation.

Regarding regex, it would be good to start by surveying existing libraries, and evaluate each of them for portability, wchar_t support, format (Posix-extended, PCRE, something else), size, license, etc. Based on   that, we can make a decision about which to go with. Maybe `regex(3)` is sufficient, I don't know!

**Comment by lloeki** [comment]
A short regex lib survey:

- [re2](https://github.com/google/re2), BSD-licensed, C++. Supports Unicode but doesn't attempt to normalise. Go uses this indirectly (has its own pure Go implementation).
- [oniguruma](http://www.geocities.jp/kosako3/oniguruma/): BSD-licensed, C. Widely used (PHP for mbyte strings, Ruby 1.9.x, text editors).
- [onigmo](https://github.com/k-takata/Onigmo): BSD-licensed, C. Forked from stagnant Oniguruma. Ruby switched from Oniguruma to this in 2.x, supports Perl 5.10+ features.

I can't comment on `wchar_t` specifically. All of them are quite portable though.

**Comment by msteed** [comment]
@ridiculousfish and anyone else who has an interest: here is a summary of what I am implementing. It follows @kballard's proposal pretty closely.
```
 string length [STRING...]
 string sub [(-s | --start)] [(-l | --length)] [STRING...]
 string split [(-n | --limit) LIMIT] SEP [STRING...]
 string join SEP [STRING...]
 string escape [STRING...]
 string match [(-a | --all)] [(-i | --ignore-case)] [(-r | --regex)]
              [(-n | --index)] [(-q | --query)] PATTERN [STRING...]
 string replace [(-a | --all)] [(-i | --ignore-case)] [(-r | --regex)]
                PATTERN REPLACEMENT [STRING...]
 # possibly add a -r|--right option to split to operate right-to-left (useful with -n)
 # other possible subcommands: upper, lower, trim, startswith, endswith
```
For documentation and examples, please take a look at https://github.com/msteed/fish-shell/blob/string-builtin/doc_src/string.txt. Thanks in advance for any feedback; I'm very open to discussion and changes.

The subcommands `length`, `sub`, `split`, `join`, `escape`, and `match` (glob matching only) are implemented. Next steps are to look into some regex engines (@lloeki, thanks for the pointers) and get some consensus there, then do regex `match`, then `replace`.

**Comment by kballard** [comment]
Regarding regex engines, IIRC, re2 doesn't support backtracking, but it also avoids the exponential-time worst case scenarios of other regex engines (which makes it safe to throw arbitrary user-supplied regular expressions at). If we're ok with saying we don't need to support backtracking then I'd be in favor of using that one.

**Comment by frangio** [comment]
I think a shell can expect to be working with reasonably small sized strings (e.g. file names), for which the backtracking penalty shouldn't be too harsh. Any script that operates with regex on long text should use a proper text processing program such as sed or grep. Furthermore, it's actually useful sometimes to have backtracking in regex, so IMO we should choose an engine capable of it.

**Comment by kballard** [comment]
Arguably, because we're likely to be working with small unstructured bits of text, it's going to be extremely rare to actually want backtracking, which means we should pick the engine with better all-around performance.

Also, I don't think it's really fair to only consider small strings (like filenames) as what the script is likely to be working with. I've had to process larger amounts of text plenty of times in bash scripts. And similarly I don't think it's fair to say "use sed or grep" because they're often not an applicable tool. For example, if I want to apply a regex substitution to every line of text, sed is great, but if I want to parse out some structured data from that text and use it in my shell script, sed is completely useless.

**Comment by ridiculousfish** [comment]
I think what's unsupported in re2 is backreferences, not backtracking.

We also ought to consider `boost::regex` too. It's header-only, supports both PCRE and POSIX, and can be easily swapped out for C++11 `<regex>` when building as C++11, since the API is nearly the same.

**Comment by kballard** [comment]
Err yeah, I meant to say backreferences. Did I really say backtracking? Oops. Regular expressions without backtracking isn't really regular expressions.

**Comment by frangio** [comment]
Yes, sorry, we were talking about backreferences all along.

@kballard makes some good points regarding parsing structured data into the shell which I hadn't considered. I think the use cases need to be sorted out before deciding on an engine.

**Comment by ridiculousfish** [comment]
@msteed  thank you for sharing your doc. It looks very good so far.

Some general feedback:
- For each command, we should determine and document what the exit status is, and leverage that as an auxiliary channel. For example, `string length` ought to return success when its input length is nonzero. In particular I hope we can provide a nicer replacement for `test -n`. 
- We should rationalize whether `string` receives its input as arguments, on stdin, or both. It's an abuse of notation but we should consider using a single dash argument to mean "read from stdin". I'm thinking especially of `join`, which we often want to stick at the end of a pipeline to get the whole output as a single argument.
- Along those lines, we should equip `string` with the superpower that it does not necessarily perform IFS splitting when used in a command substitution. In particular, it would be nice to support commands like `(cat stuff.txt | string --join \n -)` to read in the entire file as a single variable. (This idea is from my old builtin_list proposal #445.) I can take on implementing that.

**Comment by lloeki** [comment]
- split right-to-left once on `/` sounds awesome to get basename and dirname.
- piping into join is great too, helping me get rid of those `xargs echo` hacks.

**Comment by kballard** [comment]
I'm rather concerned about the idea that passing `-` should mean "read from stdin". These string operations are typically going to be run on external input, which means they should be safe to run on _any_ input. If that input happens to be `-` the script shouldn't break. You could of course define it so you'd say something like `string length -- $input` to denote that `$input` should be used as its raw value and not interpreted specially, but the problem with that is it basically means you should _always_ use `--` on every single invocation of the command, but 99% of the time people will forget or will not bother to do so. This is related to my issue with `echo`, which is that `echo` takes flags but doesn't even support `--`, so if you're echoing external input, and that input happens to begin with `-`, it can misbehave (e.g. `set -l foo -h; echo $foo`).

In fact, this a potential issue with the declared set of `string` sub-commands even without the idea of using `-` to mean stdin. Specifically the commands that take optional flags and whose first non-flag argument may be external input.

In the past I have considered the idea that built-in commands should be able to tell if the flags passed to them were written literally or came from external input (either a variable or a process substitution), so that way they can avoid misinterpreting external input as flags. I don't think I've suggested it before because it feels weird to privilege built-in commands like this in a way that external commands can't handle, but maybe that's ok. If we did do something like this, we'd need some way to annotate a variable/process substitution to indicate that it should be parsed for flags after all.

**Comment by kballard** [comment]
Looking at the list from msteed's [comment](https://github.com/fish-shell/fish-shell/issues/156#issuecomment-122483929), `string sub` has the biggest potential issue about mis-parsing input as flags. `string match` and `string replace` have an issue if the pattern was supplied by the user, but is fine otherwise (unless the pattern wants to start with a `-`). `string split` and `string join` are fine as well, assuming the separator is supplied by the script (although again if the separator wants to start with `-` that's a potential problem). But I think every single command will have a potential issue if `-` means "read from stdin".

**Comment by ridiculousfish** [comment]
@kballard The idea of distinguishing literal options from externally supplied options has been floated before, in the context of `contains`, and `echo`, and `count` which supports no options (but maybe it should).

I totally agree that if we require `--` for correctness, we've failed.

**Comment by msteed** [comment]
All the `string` subcommands, even the ones that have no options, take `--` to mean the end of option arguments, but it's true that this works only when you remember to use it. I'm not sure there is a perfect solution to that. (I'll add a note on `--` to the docs.)

I think it's okay to use `-` to represent `stdin` in place of a filename, but since `string` is not taking filenames as arguments, it doesn't quite fit. Instead, what about a command line option to take arguments from `stdin`? E.g., `(cat stuff.txt | string join \n --stdin)`

**Comment by faho** [comment]
>All the string subcommands, even the ones that have no options, take -- to mean the end of option arguments, but it's true that this works only when you remember to use it. I'm not sure there is a perfect solution to that. (I'll add a note on -- to the docs.)

The only thing I can think of is only accepting options before subcommands, but I feel dirty suggesting that.

>I think it's okay to use - to represent stdin in place of a filename, but since string is not taking filenames as arguments, it doesn't quite fit. Instead, what about a command line option to take arguments from stdin? E.g., (cat stuff.txt | string join \n --stdin)

Why have an option at all? Why not just "if something's coming in from stdin, take that, else take the arguments" - is there a usecase for mixing both?

**Comment by lloeki** [comment]
> Why not just "if something's coming in from stdin, take that"

I suppose you mean detecting if stdin is a pty, because there's always something coming in from stdin even if it's from /dev/null (where it'll wait indefinitely). It may very well be the `string` command that takes the "source" argument, so `string --stdin join` doesn't bother me that much, but that doesn't help with non-stdin, subcommand specific arguments.

To me, regarding `join` at least, the best course of action is to rely on arity: `string join ", "` will always read from stdin (similarly to grep) since it lacks a second argument. I see that some other commands can take [STRING...] but I'm not convinced on the use case nor the semantic: switching to requiring only one string instead of vararg. If one wants a vararg behavior, maybe a little polymorphism is in order and an array or strings should be used as argument. That way the last argument is always the string (or strings if that argument is an array) to operate on and the first to penultimate are subcommand flags, unless --stdin has been passed to `string` which makes the last argument omission mandatory.

Example:

```
string sub [(-s | --start)] [(-l | --length)] STRING
string --stdin sub [(-s | --start)] [(-l | --length)]
```

**Comment by msteed** [comment]
After variable expansion, how is an array of strings different from multiple string arguments?

The ambiguity around `-` at the start of an argument is bigger than `string`. How does fish deal with it elsewhere?

**Comment by ridiculousfish** [comment]
@lloeki Relying on arity is elegant but tricky, because the user may pass in an empty variable. So it would have to distinguish explicitly zero arguments from an empty argument list after expansion.

A separate possibility is for `string` to check whether its stdin is from a pipe, and read from stdin if so. The `read` builtin does this: it reads non-interactively if it's on the receiving end of a pipe. This is very simple to do.

@msteed fish generally supports `--`, and it's hard to remember to use it; this is one of the longstanding warts of Unix (`rm *` and the like). But it's especially important with `string` since `string` will get a lot of non-literal input in its argv (if you have literal input, you don't need to join or split it).

Another possible future direction is an "argument pipe", basically `xargs`:

    echo 1\n2\n3 $| string --join :

which would be syntactic sugar for

    string --join : (echo 1\n2\n3)

sort of like Haskell's `$` operator but in reverse.

**Comment by lloeki** [comment]
> @lloeki Relying on arity is elegant but tricky, because the user may pass in an empty variable. So it would have to distinguish explicitly zero arguments from an empty argument list after expansion.

Tricky? I think fish can manage this :-)

```
#!/bin/bash

function foo() {
    [[ "$#" -eq 0 ]] && echo "count zero"
    [[ -z "$1" ]] && echo "length zero"
}

echo 1
foo "42"
echo 2
foo ""
echo 3
foo
```

**Comment by ridiculousfish** [comment]
@lloeki 

Ah, but now try:

    > foo $nope
    count zero
    length zero

bash cannot distinguish between a literal-no-arguemnts and an after-expansion-no-arguments.

**Comment by lloeki** [comment]
`foo "$nope"` works though, and this is what one is supposed to write in bash (and is duly notified by the awesome `shellcheck`).

I went to bash because that's what I had right now, but I thought fish was taking care of this and that $foo and "$foo" were supposed to be the same. So you made me install fish on this machine (how sneaky of you! ;-))

#!/usr/bin/env fish

```
function foo
    test (count $argv) -eq 0; and echo "count zero"
    test -z $argv[1]; and echo "length zero"
end

echo 1
foo "42"
echo 2
foo ""
echo 3
foo
echo 4
foo $nope
echo 5
foo "$nope"
set yep ""
echo 6
foo $yep
echo 7
foo "$yep"
```

To my surprise, fish behaves like bash on test case 4. So the difference between bash and fish is only with test case 6, whereas I expected it to be different on test case 4 too. My bad. Since I only write bash scripts with `set -u`, I have the habit of only dealing with set variables, which muddled my thoughts on the issue, sorry (BTW, does fish have something like `set -u`?)

**Comment by msteed** [comment]
I am inclined to make `string` read its arguments from stdin if stdin is not a tty. That avoids the problem where `string` hangs due to an accidentally empty argv, and it is less cumbersome than requiring a switch.

Taking arguments from stdin helps a little with the ambiguity around `-`, since everything from stdin can be treated as a non-option argument. Until there is a more general solution to that problem, it seems like `--` is the best we can do.

**Comment by msteed** [comment]
Here is an update on the proposed syntax. Detailed docs including exit codes and reading strings from stdin at https://github.com/msteed/fish-shell/blob/string-builtin/doc_src/string.txt.
```
 string length [(-q | --quiet)] [STRING...]
 string sub [(-s | --start)] [(-l | --length)] [(-q | --quiet)] [STRING...]
 string split [(-n | --limit) LIMIT] [(-r | --right)] [(-q | --quiet)]
              SEP [STRING...]
 string join [(-q | --quiet)] SEP [STRING...]
 string trim [(-l | --left)] [(-r | --right)] [(-c | --chars CHARS)]
             [(-q | --quiet)] [STRING...]
 string escape [STRING...]
 string match [(-m | --max) MAX] [(-i | --ignore-case)] [(-r | --regex)]
              [(-n | --index)] [(-q | --quiet)] PATTERN [STRING...]
 string replace [(-m | --max) MAX] [(-i | --ignore-case)] [(-r | --regex)]
                [(-q | --quiet)] PATTERN REPLACEMENT [STRING...]```

**Comment by msteed** [comment]
@ridiculousfish, @lloeki, @kballard
Notes on regex engines:                                                                                             
- [re2](https://github.com/google/re2)
  - easy C++ api 
  - installed size 502KB
  - supports Perl-style [syntax](https://github.com/google/re2/wiki/Syntax) with some features omitted
  - _operates only on UTF-8 or Latin1 strings_
  - BSD license
- [pcre](http://www.pcre.org/)
  - full-featured and widely used
  - under steady development since 1997
  - installed size 3.3MB
  - has interfaces for UTF-8, UTF-16, and UTF-32 strings
  - BSD license
- [ICU](http://userguide.icu-project.org/strings/regexp)
  - full-featured and proven but big and (can be) slow
  - installed size 33.9MB (includes more than regex)
  - operates internally on UTF-16 strings but can convert to & from UTF-32
  - custom license, BSD-like
- [Oniguruma](http://www.geocities.jp/kosako3/oniguruma/)
  - deals with many encodings including UTF-16 and UTF-32
  - used in PHP for multi-byte regex
  - [sporadic updates](http://www.geocities.jp/kosako3/oniguruma/HISTORY_5X.txt)
  - installed size 543KB
  - BSD license
- [Onigmo](https://github.com/k-takata/Onigmo)
  - based on the Oniguruma engine
  - focus seems to be on Ruby support (used in Ruby 2.0)
  - less widely distributed than the others (e.g., I could not find an Arch Linux package)
  - BSD license
- [boost::regex](http://www.boost.org/doc/libs/develop/libs/regex/doc/html/index.html)
  - [relies on ICU](http://www.boost.org/doc/libs/develop/libs/regex/doc/html/boost_regex/ref/non_std_strings/icu/intro.html) for full Unicode support
  - easy to move to C++-11
  - custom license, BSD-like

All the libraries except re2 deal with catastrophic backtracking by letting the caller limit memory usage; re2 omits the features that can lead to catastrophic backtracking.

The pcre, Oniguruma, and Onigmo engines offer the best wchar_t support, and re2 offers none.

The advantage that boost offers is the mostly C++-11-compatible API. If we want to use the full-featured ICU engine, boost is the way to go. 

If size is a concern, Onigmo or Oniguruma are the clear winners, followed by pcre. (I am not considering re2 a contender because of the lack of wchar_t support.)

License is not really a differentiator.

In my opinion, pcre offers the best balance of Unicode friendliness, rich features, wide availability, and long-term viability.

**Comment by zanchey** [comment]
The only problem I can see with pcre2 is that it is so new that it's not available on many of the platforms we support - RHEL/CentOS 5/6/7, Debian 6/7/8, Fedora 21/22, openSUSE 13.1/13.2, or OS X (sans Homebrew or other third-party packages). Mind you, OS X doesn't have any of the options listed, so fish will need to vendor the source for at least some configurations.

pcre2 is available in Homebrew and Fedora 23.

**Comment by lloeki** [comment]
My understanding was that things were learning more towards vendoring the dependency so that it gets built alongside fish (whether it is by source or by way of git submodules, that I don't venture to say).

**Comment by ridiculousfish** [comment]
What does "vendoring the dependency mean?" I had assumed we would integrate the source directly but maybe that was baseless.

**Comment by lloeki** [comment]
> What does "vendoring the dependency mean?"

Include (a version of) the dependency's source code into the SCM alongside your own, typically in a dedicated directory. I don't know if the name comes from there but Rails conventionally uses a `vendor` directory for that, which was used to store non-app assets and gems before the advent of rubygems.org and bundler.

**Comment by msteed** [comment]
I am almost ready to start on the regex functionality. Any objection to using pcre2 then?

**Comment by ridiculousfish** [comment]
Go for it. We should measure the binary size before and after though and make sure that we're getting stripping and dead code elimination like we expect.

**Comment by msteed** [comment]
After adding the pcre2 calls to do regex string matching, plus the code for the other string subcommands, the increase in the size of the fish binary is 178KB (x86/64). Is that okay?

**Comment by ridiculousfish** [comment]
Yes, seems OK. I and others can put it on a diet :grinning:

**Comment by msteed** [comment]
Does fish make any assumptions about `sizeof(wchar_t)`? I'm looking at whether I need to make the pcre2 calls work with both 16-bit and 32-bit wide characters.

**Comment by ridiculousfish** [comment]
fish assumes wchar_t is 32 bits (*nix) or 16 bits (Windows).

**Comment by msteed** [comment]
Regex `string match` is working, leaving only regex `string replace` to be done.

A couple outstanding questions:

0. What should `string split` do with an empty separator? Possibilities: do no splitting, split individual characters, or fail. Currently it does no splitting, but I'm inclined to change it to split individual characters.

1. In `configure.ac` there is a check for `regex.h` but that header is not used anywhere. Should I remove the check?

**Comment by ridiculousfish** [comment]
1. Split to all individual characters is right

2. Yes, thanks!

**Comment by faho** [comment]
This has been fixed via #2296, no?

**Comment by ridiculousfish** [comment]
woohoo!
